[FEAT/#75] login / firebase authentication 도입

shinythinking opened the pull request:
<details>
  <summary>[FEAT/#75]</summary>

- closed #75 

## *📍 Work Description*

- firebase authentication 관련 의존성 추가
- firebase authentication 도입하여 구글 소셜 로그인 구현
- 멀티 모듈 정리
- ui에 반영

## *📸 Screenshot*
(<img src="https://github.com/user-attachments/assets/6d24abe7-321c-42c4-979a-9139719dc309" width=270 />)


## *📢 To Reviewers*
- firebase authentication을 이용해서 구글 소셜 로그인을 구현했습니다.
- 처음 해보는 작업인지라 많은 우여곡절을 겪었습니다.
  - firebase 공식문서에는 xml 코드만 있다!!
  - 있는 코드도 Deprecate된 함수들을 사용한다...
  - 안드로이드 공식 문서에 해답이 있었다. 다만 사용법만 있어서 멀티모듈로 변환하는 작업이 필요했다.
  - hilt관련 문제로 5시간은 씨름한 것 같다. -> app 모듈에 의존성을 추가하여 해결!
- MVI로 바꾸는 작업을 추후에 진행하겠습니다! 
- 위의 스크린샷은 연출된 장면입니다! 이 브랜치를 받아서 돌리면 이 화면이 나오지 않습니다. 
- 버전별 대응을 해야할 것 같습니다.
- LocalProperty에 포함될 값 `비밀이야기`에 추가했습니다.

## ⏲️Time

    - 12 h +
</details>

<details>
  <summary>[FIX/#84]</summary>

- closed #84

## *📍 Work Description*
- data:auth, domain:auth -> core:auth 이전
- 의존성 설정
- google auth 와 firebase auth 를 통해 googleIdToken, UserInfo, firebaseUID 가져옴
- Activity Context를 사용하여 API30~33 에서 구글 로그인이 동작을 안하던 문제 해결

## *📸 Screenshot*

https://github.com/user-attachments/assets/a947ffa5-35db-4a20-8182-1f27db3ec011

<img src="https://github.com/user-attachments/assets/62f12e1c-9bc7-4e0c-b4bd-a9bb866e40bd" width=500>

## *📢 To Reviewers*
- 현재 로그인 화면에서는 어떠한 정보도 저장되지 않아 로그인 성공시 Toast 화면에 유저 정보만 나타나도록 했습니다.
- 구글 로그인 로직이 context와 얽혀있을 뿐만 아니라 Activity Context를 요구하기 때문에 hilt 는 제외했습니다.
- 동작 순차
  - 1. google cloud authentication 에서 유저 인증을 요청하고 Credential를 받습니다.
  - 2. Credential 이 올바른지 검사합니다. 올바르다면 GoogleIdTokenCredential 으로 타입캐스팅 됩니다.
  - 3. GoogleIdTokenCredential 의 idToken를 사용하여 firebase에 로그인을 요청합니다. 요청의 성공/실패 여부에 따라 개인 정보를 얻을 수 있습니다.
  - 4. 성공시 GoogleIdTokenCredential 에서 유저 정보(idToken 포함)와 firebaseUser의 UID를 뷰모델에 전달합니다.

- idToken : firebase 에 로그인 인증을 하기 위해 구글에서 발급한 토큰 (JWT)
- UID : firebase에서 유저를 구분하기 위해 발급된 시리얼(영구)

- 현재 구글 계정에서 임시로 정보를 추출했습니다.
  - 성, 이름
  - 휴대폰 번호
  - idToken
- 이것 외에도 아래와 같이 몇 가지 정보를 더 제공해줍니다.
<img src="https://github.com/user-attachments/assets/f1991b4f-a008-4f04-bded-fe334bd06378" width=500>

- local.properties 에 GOOGLE_WEB_CLIENT_ID 값을 별도로 추가해야 빌드가 됩니다. (local.properties gitignore 때문에 lint 빌드 실패)

## ⏲️Time

    - 5 h

</details>

TaewoongR committed:
feat/#75 : 구글 oauth 의존성 추가

- com.google.android.gms:play-services-auth
- com.google.android.libraries.identity.googleid:googleid
- androidx.credentials:credentials

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : 구글 아이콘 추가

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : 임시 로그인 화면 생성

- 임시로 Material 화면으로 구성

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : Google Cloud Oauth Web Client Id 추가

- local.property에 값을 저장하고 BuildConfig를 통해 얻도록 설정

Co-authored-by: shinythinking <shinythink@gmail.com>

TaewoongR committed:
feat/#75 : Google Oauth 로직 설정

- 뷰모델에서 로직 설계
- 이를 domain, data 모듈로 이전 예정

Co-authored-by: shinythinking <shinythink@gmail.com>

shinythinking committed:
chore/#75 : credential 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 도메인 모듈 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, User 모델 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, SignInWithGoogleUseCase 구현

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 도메인 auth, AuthRepository 구현
* 현재는 로그인 함수만 존재
* 로그아웃은 구현 예정

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 데이터 auth, 의존성 추가
* Local에서 Client에 대한 값을 받아오는 로직 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, GoogleAuthDataSource 인터페이스 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, GoogleAuthDataSource 구현체 추가
* application context를 사용해서 credential manager를 사용
* setNounce를 사용하여 보안 강화

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, FirebaseAuthDataSource 인터페이스 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, FirebaseAuthDataSource 구현체 추가
* firebaseAuth를 이용해서 인증 수행

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, mapper 클래스 추가
* 도메인의 User와 AuthResult와 매핑

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, AuthRepository 구현체 추가
* googleAuthDataSource와 firebaseAuthDataSource, UserMapper를 hilt로 받아 사용

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 데이터 auth, di 추가
* FirebaseAuth
* GoogleAuthDataSource
* UserMapper
* AuthRepository
* 의 의존성 주입!

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 피처 auth, 의존성 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : 임시 화면(시험용) 삭제

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, AuthUiState 설정
* 성공시, User를 인수로 받도록 설정

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, AuthViewModel 구현
* Usecase를 받아서 구글 로그인 구현

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, di 설정
* SignInWithGoogleUseCase 의존성 주입!

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, LoginScreen 구글 로그인 버튼 추가

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
feat/#75 : 피처 auth, LoginScreenRoute 추가
* Route에서 uiState 변화를 캐치

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
chore/#75 : app 수준에 의존성 추가
* (for Hilt!)

Co-authored-by: TaewoongR <iwoong98@kau.kr>

shinythinking committed:
fix/#75 : ktlint의 요구사항...

Co-authored-by: TaewoongR <iwoong98@kau.kr>

Ameri-Kano commented on app/build.gradle.kts:
```diff
@@ -37,7 +37,13 @@ dependencies {
 	implementation(platform(libs.firebase.bom))
 	implementation(libs.bundles.firebase)
-	implementation(projects.feature.main)
+	implementation(libs.google.play.services.auth)
+	implementation(libs.googleid.identity)
+	implementation(libs.androidx.credential)
```
이쪽을 auth 관련 bundles 로 묶는 것도 좋아 보입니다.

shinythink replied:
좋은 의견 감사합니다!!

Ameri-Kano commented on data/auth/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest>
+
+</manifest>
```
필요없다면 삭제해도 좋을 것 같네요!

shinythink replied:
commit에 넣지 않으려고 했는데 들어갔네요. 나중 Commit에서 삭제하겠습니다.

Ameri-Kano commented on data/auth/src/main/java/com/boostcamp/mapisode/auth/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,27 @@
+	private val firebaseAuthDataSource: FirebaseAuthDataSource,
+	private val userMapper: UserMapper,
+) : AuthRepository {
+	override suspend fun signInWithGoogle(): Flow<Result<User>> = callbackFlow {
+		try {
+			val googleCredential = googleAuthDataSource.getGoogleCredential()
+			val firebaseAuthResult = firebaseAuthDataSource.signInWithGoogle(googleCredential)
+			val user = userMapper.mapToUser(firebaseAuthResult)
+			trySend(Result.success(user))
+		} catch (e: GetCredentialCancellationException) {
+			trySend(Result.failure(e))
+		} catch (e: Exception) {
+			trySend(Result.failure(e))
+		}
+		awaitClose { }
+	}
```
callbackFlow 는 일반적인 flow 와 어떤 차이가 있는 건가요?

shinythink replied:
Firebase나 Google Auth 같은 콜백 기반 API를 Flow로 변환할 때 사용할 수 있습니다!
* trySend()로 데이터 방출하고
* awaitClose()로 리소스 정리합니다.

Ameri-Kano commented on feature/login/src/main/java/com/boostcamp/mapisode/login/AuthViewModel.kt:
```diff
@@ -0,0 +1,37 @@
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class AuthViewModel @Inject constructor(
+	private val signInWithGoogleUseCase: SignInWithGoogleUseCase,
+) : ViewModel() {
+
+	private val _uiState = MutableStateFlow<AuthUiState>(AuthUiState.Initial)
+	val uiState = _uiState.asStateFlow()
+
+	fun handleGoogleSignIn() {
+		viewModelScope.launch {
+			_uiState.value = AuthUiState.Loading
+			signInWithGoogleUseCase().collect { result ->
+				_uiState.value = when {
+					result.isSuccess -> {
+						val user = result.getOrNull()
+						AuthUiState.Success(user)
+					}
+
+					else -> {
+						AuthUiState.Error("Sign-in failed")
+					}
+				}
+			}
+		}
+	}
+}
```
MVI가 적용된 것 같지는 않네요! 그랬다면 시간이 너무 많이 걸렸겠죠?

shinythink replied:
네 ㅠㅠ 우선 익숙한 걸로 구현했습니다. 추후에 MVI 적용할 생각입니다!

Ameri-Kano reviewed (approved):
고생 많으셨습니다! 굉장히 소중한 경험이라고 생각해서, 꼭 문서로 남겨두었으면 좋겠습니다.

shinythinking committed:
fix/#75 : ktlint의 요구사항2...

Co-authored-by: TaewoongR <iwoong98@kau.kr>

haeti-dev commented on data/auth/src/main/java/com/boostcamp/mapisode/auth/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,27 @@
+import kotlinx.coroutines.flow.callbackFlow
+import javax.inject.Inject
+
+class AuthRepositoryImpl @Inject constructor(
+	private val googleAuthDataSource: GoogleAuthDataSource,
+	private val firebaseAuthDataSource: FirebaseAuthDataSource,
+	private val userMapper: UserMapper,
+) : AuthRepository {
```
q) 구현체의 코드를 직접 삽입 안하고 인터페이스를 사용한 이유가 있나요? 인터페이스를 꼭 분리했어야 했나요?

shinythink replied:
다른 로그인이 추가될 가능성을 염두에 두었습니다. 이렇게 하면 다른 소셜 로그인과 관련된 코드를 짜고 기존 코드를 확장만 하면 되기 때문입니다.

haeti-dev replied:
그 답변도 설득력 있긴 하지만, 사실 테스트 관련 답변을 기대했습니다 ㅎ

haeti-dev commented on domain/auth/src/main/java/com/boostcamp/mapisode/auth/User.kt:
```diff
@@ -0,0 +1,3 @@
+package com.boostcamp.mapisode.auth
+
+data class User(val id: String, val displayName: String?)
```
core:model에 있어도 되지 않을까요? 옮기면 나중에 presentation 레이어에서 편하게 사용할 수 있을 듯 합니다.

shinythink replied:
동의합니다! 옮기도록 하겠습니다!

haeti-dev commented on app/build.gradle.kts:
```diff
@@ -37,7 +37,13 @@ dependencies {
 	implementation(projects.core.ui)
 	implementation(projects.core.designsystem)
 	implementation(projects.core.navigation)
+	implementation(projects.domain.auth)
+	implementation(projects.data.auth)
+	implementation(projects.feature.login)
+	implementation(projects.feature.main)
```
이 모듈들이 app 모듈에 추가되어야 하는 이유는 뭔가요?

shinythink replied:
전이 의존성 때문에 data:auth만 추가해도 동작은 합니다. 다만 의존 관계를 명확히 보여주고자 다 추가했습니다. 다만, 다른 의존성들이 여기에 추가될 것을 생각하면 data:auth만 추가해도 괜찮을 것 같습니다. 어떻게 하는 게 좋을까요? 의견이 궁금합니다!
ps) 좀 더 공부를 해야할 것 같습니다 ㅠ

haeti-dev replied:
의존 관계를 명확히 보여주고자 다 추가했다는게 좀 말이 안되네요..? 당연히 필요한 것만 의존해야됩니다. 순환 의존성이 발생할 확률을 줄여주세요. 또 빌드 시간에 대한 이점을 가지는 멀티모듈 구조를 잘 활용하기 위해서는 줄여야 합니다

haeti-dev reviewed (changes requested):
고생하셨습니다 리뷰 확인 부탁드려요~
모듈 간의 결합도를 낮추는 게 중요한 것 같아서, login 모듈 관련해서는 수정이 필요할 것 같습니다.

shinythinking committed:
refactor/#75 : bundle.auth로 의존성 묶기

shinythinking committed:
refactor/#75 : User 데이터 클래스를 core 모듈로 이동

TaewoongR commented on data/auth/src/main/java/com/boostcamp/mapisode/auth/GoogleAuthDataSourceImpl.kt:
```diff
@@ -0,0 +1,53 @@
+				}
+			}
+
+			else -> throw RuntimeException("Invalid credential")
```
어제 이 부분에 대해서 문제가 있었음을 확인했습니다. 
사인님은 알고 계시겠지만, API 35 기기에서는 실행이 되지만 API 33 기기에서는 로그인 화면 창이 나타나지 않습니다.
<img src="https://github.com/user-attachments/assets/bc18d403-80df-4f32-b50c-3fcb9ac34610" width=800>
이를 해결하기 위해 ActivityContext를 주입하려 했지만, 진행하면서 되돌아보니 domain, data 모듈에서는 생명주기를 알고 있어야 하는가? 의 질문에 ui layer 외에는 의존하면 안된다 라는 결론을 세우고 리팩토링 과정을 중단했습니다.

또한, ActivityContext를 주입을 할 수 없었던 이유는 domain 모듈은 android 의존성이 없기 때문에 Activity 주기를 모릅니다.
그래서 data 모듈의 di module에서 스코프를 지정하는 과정에서 ActivityComponent로 설정할 때 domain 모듈에게 주입하는 과정에서 오류가 발생했습니다.

현재 제 생각은 이 로직 전부가 data 모듈이 아닌 feature:login 모듈에 존재하면 안될까 라는 의견을 갖고 있습니다.

TaewoongR reviewed (commented):
data, domain 모듈을 사용하지 않고 feature:login 에서 전부 정의하면 어떨까란 의견입니다.

shinythinking committed:
refactor/#75 : 로그인 모듈 의존성 변경
* feature:login, domain:auth 제거했습니다.

TaewoongR committed:
fix/#84 : core:auth 모듈 생성

TaewoongR committed:
fix/#84 : google oauth 로직 위치 이전

TaewoongR committed:
fix/#84 : google oauth 결과 데이터 클래스 타입 선언

- 구글 계정에서 받은 실명과 휴대폰 번호
- 자동 로그인을 위한 google idToken
- firebase에서 받은 유저 구분을 위한 UID

TaewoongR committed:
fix/#84 : 모듈 삭제

- 로직 이전 후 필요 없는 모듈 일괄 삭제 및 의존성 수정

TaewoongR committed:
fix/#84 : 추가적인 미사용 파일 삭제

TaewoongR committed:
fix/#84 : 의존성 수정

- feature:login 에서 auth 관련 의존성 삭제

TaewoongR committed:
fix/#84 : core:auth 와 feature:login 연결

- GoogleOauth - ViewModel- Screen 연결
- Flow의 callbackFlow 내부에서 trySend 를 반환해야 collect 호출 시 emit

TaewoongR committed:
fix/#84 : fix for ktlint

TaewoongR committed:
fix/#84 : fix for ktlint
